src/main/java/com/strandls/observation/service/Impl/ObservationListServiceImpl.java

```diff
@@ -511,6 +526,161 @@ public MapAggregationResponse mapAggregate(String index, String type, String sGr
+
+		String omiter = null;
+		MapAggregationStatsResponse aggregationStatsResponse = new MapAggregationStatsResponse();
+
+		Map<String, AggregationResponse> mapAggStatsResponse = new HashMap<String, AggregationResponse>();
+
+		int totalLatch = 2;
+//		latch count down
+		CountDownLatch latch = new CountDownLatch(totalLatch);
+
+		if (recoName != null && !recoName.isEmpty()) {
+			mapSearchQueryFilter = esUtility.getMapSearchQuery(sGroup, taxon, user, userGroupList, webaddress,
+					speciesName, mediaFilter, months, isFlagged, minDate, maxDate, validate, traitParams, customParams,
+					classificationid, mapSearchParams, maxvotedrecoid, recoId, createdOnMaxDate, createdOnMinDate,
+					status, taxonId, omiter, rank, tahsil, district, state, tags, publicationGrade, authorVoted);
+
+			getAggregateLatch(index, type, "max_voted_reco.scientific_name.keyword", geoAggregationField,
+					mapSearchQueryFilter, mapAggStatsResponse, latch, null);
+
+		} else {
+
+			getAggregateLatch(index, type, "max_voted_reco.scientific_name.keyword", geoAggregationField,
+					mapSearchQuery, mapAggStatsResponse, latch, null);
+		}
+
+		// for top Uploaders
+
+		if (user != null && !user.isEmpty()) {
+			mapSearchQueryFilter = esUtility.getMapSearchQuery(sGroup, taxon, omiter, userGroupList, webaddress,
+					speciesName, mediaFilter, months, isFlagged, minDate, maxDate, validate, traitParams, customParams,
+					classificationid, mapSearchParams, maxvotedrecoid, recoId, createdOnMaxDate, createdOnMinDate,
+					status, taxonId, recoName, rank, tahsil, district, state, tags, publicationGrade, authorVoted);
+
+			getAggregateLatch(index, type, "author_id", geoAggregationField, mapSearchQueryFilter, mapAggStatsResponse,
+					latch, null);
+
+		} else {
+
+			getAggregateLatch(index, type, "author_id", geoAggregationField, mapSearchQuery, mapAggStatsResponse, latch,
+					null);
+		}
+
+		try {
+			latch.await();
+		} catch (Exception e) {
+			logger.error(e.getMessage());
+		}
+
+		int size = lifeListOffset + 10;
+		int count = 1;
+
+		Map<String, Long> temp = mapAggStatsResponse.get("max_voted_reco.scientific_name.keyword")
+				.getGroupAggregation();
+
+		Map<String, Long> t = new LinkedHashMap<>();
+
+		if (recoName != null && !recoName.isEmpty()) {
+			t.put(recoName, temp.get(recoName));
+
+			aggregationStatsResponse.setGroupUniqueSpecies(t);
+
+		} else {
+
+			for (Map.Entry<String, Long> entry : temp.entrySet()) {
+
+				if (count <= (size - 10)) {
+					count++;
+				} else {
+					if (count > size) {
+						break;
+					}
+					t.put(entry.getKey(), entry.getValue());
+					count++;
+				}
+			}
+			aggregationStatsResponse.setGroupUniqueSpecies(t);
+
+		}
+		int uploadersSize = uploadersoffset + 10;
+		int uploadersCount = 1;
+		String authorIds = "";
+
+		List<Long> counts = new ArrayList<>();
+
+		Map<String, Long> uploaders = mapAggStatsResponse.get("author_id").getGroupAggregation();
+
+		if (user != null && !user.isEmpty()) {
```

Comment:
try to write extraction logic in  private function and  call it when req, instead writing inside this block

Reply:
added a private function for extraction logic